File: Algorithm/src/LinkedList/LeetCode19.java
package LinkedList;

import java.util.Stack;

public class LeetCode19 {

    class  ListNode{
        int val;
        ListNode next;
        ListNode(){};
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null){
            return  null;
        }

        ListNode dumm = new ListNode(0,head); // 设置一个哑节点
        Stack<ListNode> stack = new Stack<>();
        ListNode cur = dumm;

        while(cur != null){
            stack.push(cur);
            cur = cur.next;
        }

        for(int i  = 0; i < n; i++){
            stack.pop();
        }
        ListNode prev = stack.peek(); //栈顶节点
        prev.next = prev.next.next;
        ListNode ans = dumm.next;
        return ans;
    }
}

File: top-interview-questions-easy/LinkedList/LeetCode19.go
package LinkedList

type ListNode struct {
	Val  int
	Next *listNode
}

func getListLength(head *ListNode) int {
	nodeNum := 0
	for head != nil {
		nodeNum++
		head = (*ListNode)(head.Next)
	}
	return nodeNum
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {

	listLength := getListLength(head) // 获取链表长度
	dummy := &ListNode{0, head}       //哑节点，通过哑节点处理当前头节点特殊情况
	curr := dummy
	for i := 0; i < listLength-n+1; i++ {
		curr = curr.Next
	}
	curr.Next = curr.Next.Next
	return dummy.Next
}
